---
title: "Infer ASVs with DADA2"
author: "Calla Bush St George"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.align = "center",
                      fig.path = "../figures/01_DADA2/") #Send any figure output to this folder
```

# Goals for this file

1. Use raw fastq and generate the quality plots to asses the quality of reads

2. Filter and trim out bad sequences and bases from our sequencing files

3. Write out fastq files with high quality sequences

4. Evaluate the quality from our filter and trim. 

#Load Libraries
```{r load-libraries}
#install.packages("devtools")
library("devtools")

#devtools::install_github("benjjneb/dada2")
library(dada2)
```

# Load Data
```{r load data}
#Set the raw fastq path to the raw sequencing files
#Path to the fastq files
raw_fastqs_path <- "data/01_DADA2/01_raw_gzipped_fastqs"

#What files are in this path (Intuition check)
list.files(raw_fastqs_path)

#How many files are there?
str(list.files(raw_fastqs_path))

#Create a vector of forward reads
forward_reads <- list.files(raw_fastqs_path, pattern = "R1_001.fastq.gz", full.names = TRUE) 
#Intuition check
head(forward_reads)

#Create a vector of reverse reads
reverse_reads <-list.files(raw_fastqs_path, pattern = "R2_001.fastq.gz", full.names = TRUE)
#Intuition check
head(reverse_reads)

```

# Quality plots
```{r raw-quality plots}
#Randomly select 2 samples from dataset to evaluate
random_samples <- sample(1:length(forward_reads), size =2)
random_samples

#Calculate and plot quality of these two samples
plotQualityProfile(forward_reads[random_samples])
plotQualityProfile(reverse_reads[random_samples])
```

# Prepare a placeholder for filtered reads
```{r prep-filtered-reads}
# vector of our samples, extract the sample information from our file
samples <- sapply(strsplit(basename(forward_reads), "_"), `[`,1)
#Intuition check
head(samples)
#place filtered reads into filtered_fastqs_path
filtered_fastqs_path <- "data/01_DADA2/2_filtered_fastqs"

# create 2 variables : filtered_F, filtered_R
filtered_forward_reads <- file.path(filtered_fastqs_path, paste0(samples, "_R1_filtered.fastq.gz"))
#Intuition check
length(filtered_forward_reads)


filtered_reverse_reads <- file.path(filtered_fastqs_path, paste0(samples, "_R2_filtered.fastq.gz"))
#Intuition check
length(filtered_reverse_reads)
```


