---
title: "Phyloseq PreProcessing"
author: "Calla Bush"
date: "`r Sys.Date()`"
output:
  html_document: 
    code_folding: show
    theme: spacelab
    highlight: pygments
    keep_md: no
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: yes
      toc_depth: 3
  keep_md: true  
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.align = "center",
                      fig.path = "../figures/02_PreProcessing/")
# Send any figure output to the 02_PreProcessing folder
```


# Goals

Here, we will process the data into a phyloseq object.

- ASV table
- Taxonomy table
- Track Reads (metadata)

Then, we will remove the following:

1. Remove Chloroplasts.
2. Remove Mitochondria.
3. Remove ASVs from negative controls and alos negative control samples.
4. Evaluate accuracy by looking at the Mock Community
5. Remove samples without "enough" reads.


Finally, write a data file of phyloseq output, which will be stored in `data/02_PreProcessing/raw_preprocessed_phyloseq.RData`.

# Load libraries
```{r load-libraries}

#install_github("joey711/phyloseq")

pacman::p_load(devtools, phyloseq, tidyverse, dada2, install = FALSE)
```


# Load data

## ASV table
```{r load-ASV-table}
# First, load ASV table
load("data/01_DADA2/ASV_counts.RData")

# Inspect ASV_tab
head(asv_tab)[,1:5]

# Fix names
sample_names <- colnames(asv_tab)
samples_fixed <- sapply(strsplit(basename(sample_names), "_"), `[`,1)
head(samples_fixed)

# re-write the ASV count file to fix names
colnames(asv_tab) <- samples_fixed
str(asv_tab)
```

## Taxonomy table
```{r load-tax-table}
tax_df <- read.table("data/01_DADA2/ASV_taxonomy.tsv", sep = "\t", skip = 1)
head(tax_df)

# fix column names
colnames(tax_df) <- c("asv_names", "Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species", "ASV", "ASVseq")

head(tax_df)

#Taxonomy Table Matrix
tax_mat <-
tax_df %>%
  tibble::column_to_rownames(., var = "asv_names") %>%
  as.matrix()
```

## Track Reads Data
```{r load-track-reads}
load("data/01_DADA2/track_read_counts.RData")

# Take a look at the data
track_counts_df

# Make sure all data in ASV table matches ASV taxonomy
head(track_counts_df)
dim(track_counts_df)

# Please in the terminal copy metadata.csv to data folder 
# into: /workdir/cab565/git_repos/SalinityGradient_16S/data
# from: /workdir/in_class_data/SalinityGradient_16S/metadata.csv

#Load in metadata
metadata_df <- read.csv("data/metadata.csv")
head(metadata_df)
dim(metadata_df)

# Merge metadata df with track_reads_df
head(metadata_df)
head(track_counts_df)

metadata_track_reads_df <-
  metadata_df %>% 
  left_join(., track_counts_df, by = "names") %>%
  dplyr::select(-X)

head(metadata_track_reads_df)

# Update row.names to be sample names
## Before
row.names(metadata_track_reads_df)
## Rewrite
row.names(metadata_track_reads_df) <- metadata_track_reads_df$names
## Check that it worked
row.names(metadata_track_reads_df)
# Intuition check
head(metadata_track_reads_df)
```

# Handoff to phyloseq
```{r phyloseq-handoff}
# Double check it's all good
dim(asv_tab)
dim(tax_mat)

# Intuition check
stopifnot(row.names(asv_tab) == row.names(tax_mat))

# Construct the phyloseq object
raw_physeq <- phyloseq(otu_table(asv_tab, taxa_are_rows = TRUE), 
                       sample_data(metadata_track_reads_df),
                       tax_table(tax_mat))
raw_physeq

# Save this raw phyloseq object
save(raw_physeq, file = "data/02_PreProcessing/raw_physeq.RData")
```

# Clean up the data

Remove: 

1. Chloroplasts
2. Mitochondria

```{r rm-mitos-chloros}
#Remind myself of the tax matrix
#View(tax_mat)

# Make a new physeq without chloroplasts
noChloros_physeq <-
  raw_physeq %>%
  # rm chloroplasts
  subset_taxa(Order != "Chloroplast" | is.na(Order))

# How many taxa were chloroplasts
num_chloro_ASVs <- ntaxa(raw_physeq) - ntaxa(noChloros_physeq)
num_chloro_ASVs

#Intuition check
#noChloros_physeq %>%
#  tax_table() %>%
#  data.frame() %>%
#  View()

#Remove mitochondria
noChlorosMitos_physeq <-
  noChloros_physeq %>%
  # rm chloroplasts
  subset_taxa(Family != "Mitochondria" | is.na(Family))

# How many mitochondrial ASV?

num_mito_ASVs <- ntaxa(noChloros_physeq) - ntaxa(noChlorosMitos_physeq)
num_mito_ASVs

noChlorosMitos_physeq

# How many total ASVs were removed from Chloros and Mitos?
ntaxa(raw_physeq) - ntaxa(noChlorosMitos_physeq)
# Proportion of ASVs?
ntaxa(noChlorosMitos_physeq)/ntaxa(raw_physeq)
```

# Evaluate and remove the control samples

Take a look at the negative controls and then make a decision about whether or not to remove the ASVs that we found in our controls

1. Negative controls
2. ASVs found within the negative controls and their distribution in the samples. 
3. Evaluate the mock community

## Negative Controls
```{r neg-controls}
# Create a vector of samples that are negative controls
control_samples <- c("WaterControl", "022um-Control", 
                     "3um-Control", "DNA-Ext-Control")

# Make a new phyloseq object of only the control samples
control_physeq <-
  noChlorosMitos_physeq %>%
  subset_samples(., names %in% control_samples) %>%
  # Force remove ASVs with 0 counts
  prune_taxa(taxa_sums(.) > 0, .)

control_physeq

# Vector of control ASVs
control_ASVs <-
  control_physeq %>%
  tax_table() %>%
  data.frame() %>%
  dplyr::select(ASV) %>%
  as.vector()

# Visualize raw abundance values of control ASVs
control_physeq %>%
  plot_bar(., "ASV", fill = "ASV")

################# Check the control ASVs in the samples
# Make a new phyloseq object with only the control ASVs and all samples
control_ASV_physeq <-
  noChlorosMitos_physeq %>%
  subset_taxa(., ASV %in% control_ASVs$ASV)

# Inspect it
control_ASV_physeq

# Let's take a look at the abundances
control_ASV_physeq %>%
  plot_bar(., "ASV", fill = "ASV", facet_grid = Sample_or_Control~.)

#Check tax of ASV 3
#control_ASV_physeq %>% 
#  tax_table() %>%
#  data.frame() %>%
#  View()

# Would only like to keep ASV 18 and ASV 56. Everything else is trash.
control_ASVs_toREMOVE <- dplyr::setdiff(control_ASVs$ASV,
                                        c("ASV_18", "ASV_56"))

# Intution check 
control_ASVs_toREMOVE
length(control_ASVs_toREMOVE)

### Make new phyloseq object
noControlsChlorosMitos_physeq <-
  noChlorosMitos_physeq %>%
  # Actually remove the control ASVs except 18 and 56
  subset_taxa(., !(ASV %in% control_ASVs_toREMOVE)) %>%
  # Remove control samples
  subset_samples(., !(names %in% control_samples))

# Inspect no controls object
noControlsChlorosMitos_physeq

# After filtering, how many ASVs were removed?
total_ASVs_rm <- ntaxa(raw_physeq) - ntaxa(noControlsChlorosMitos_physeq)

```

# Evaluate Accuracy in the Mock Community

- Make a mock phyloseq object
- Load in the standard fasta sequences of the mock community
- Test for differences between the two

```{r mock_physeq}
# Make mock phyloseq object
mock_physeq <-
  noControlsChlorosMitos_physeq %>%
  subset_samples(., names == "MockZymoPos") %>%
  # Mock zymo community is our positive control 
  prune_taxa(taxa_sums(.) > 0, .) 

# Inspect it
mock_physeq

tax_table(mock_physeq)

# Write us a message summarizing this
cat("DADA2 inferred", ntaxa(mock_physeq), "ASVs present in the mock community.")

######## Load in mock reference sequences from Zymo
mock_ref <- getSequences("data/02_PreProcessing/mock_amplicons.fasta")
names(mock_ref)

### Pull out seqs from Sequenced Mock community
mock_seqs <-
   mock_physeq %>%
     tax_table() %>%
     data.frame %>%
     dplyr::select(ASV, ASVseq)
# Inspect
head(mock_seqs)

# Which ASVs match the reference?
matched_asvs <-
  mock_seqs %>%
  rowwise() %>%
  # Check each ASV if they match, will equal TRUE, if not = FALSE
  mutate(Match = any(grepl(ASVseq, mock_ref))) %>%
  # Create a vector of names that matched
  pull(ASV)

cat(length(matched_asvs), "ASVs were exact matched to the expected reference sequence." )

# Evaluate the sequence depth by looking at mock_physeq
mock_SeqDepth <-
  mock_physeq %>%
  otu_table() %>%
  data.frame() %>%
  colSums()

# Evaluate which ASVs do not match
# What are their abundances
# This helps provide error rate

# Who is in the sequenced mock community?
mock_physeq %>%
  tax_table() %>%
  data.frame() %>%
  dplyr::select(Genus, Species)

names(mock_ref)

# Curious: What are the mock counts in the actual samples
noControlsChlorosMitos_physeq %>%
  subset_taxa(., ASV %in% matched_asvs) %>%
  otu_table() %>%
  t()
```

What can we conclude about the mock community? That all 8 ASVs in the mock exactly match 8 of 9 in the reference file. 

```{r rm-mock-asvs}
# Make a new phyloseq object without mock and its asvs
noControlsMockChlorosMitos_physeq <-
  noControlsChlorosMitos_physeq %>%
  # remove the mock community
  subset_samples(., names != "MockZymoPos") %>%
  # remove the ASVs from the mock
  subset_taxa(., !(ASV %in% matched_asvs))

# Intuition check
num_ASVs_mock <- ntaxa(noControlsChlorosMitos_physeq) -
  ntaxa(noControlsMockChlorosMitos_physeq)

stopifnot(num_ASVs_mock == 8)
```

# Evaluate the Sequencing Depth 

```{r seq-depth}
# The current data object
noControlsMockChlorosMitos_physeq

# What is the library size/sequencing depth for each sample? 
seqSums_df <- 
  noControlsMockChlorosMitos_physeq %>%
  otu_table() %>%
  # Sum each sample column 
  colSums() %>%
  data.frame() %>%
  rownames_to_column(var = "names") %>%
  left_join(., metadata_track_reads_df, by = "names") 

# Rename second column 
colnames(seqSums_df)[2] <- "TotalSeqs"

# check
dim(seqSums_df)
head(seqSums_df)

# Show the depth of samples 
seqSums_df %>%
  dplyr::select(names, TotalSeqs) %>%
  arrange(TotalSeqs) %>%
  head()

# plot it! 
seqSums_df %>%
  ggplot(aes(x=reorder(names, TotalSeqs), y = TotalSeqs,
             fill = station)) + 
  geom_bar(stat = "identity") 

# Density plot 
seqSums_df %>%
  ggplot(aes(TotalSeqs, fill = station)) +
  geom_density(alpha = 0.5)
```

# Remove samples with few reads 

```{r rm-samps}
# What's the min seq depth? 
min(sample_sums(noControlsMockChlorosMitos_physeq))

# Remove 20210615-MA-SCB2F
raw_preprocessed_physeq <- 
  noControlsMockChlorosMitos_physeq %>%
  subset_samples(., names != "20210615-MA-SCB2F")

#What's the new min seq depth?
min(sample_sums(raw_preprocessed_physeq))
```

# Save Preprocessed Phyloseq Object
```{r save-physeq}
save(raw_preprocessed_physeq,
     file = "data/02_PreProcessing/raw_preprocessed_physeq.RData")
```


##Session information
```{r session-info}
#Ensure reproducibility
devtools::session_info()
```




